src/main/java/at/ac/tuwien/kr/alpha/grounder/Substitution.java

```diff
@@ -0,0 +1,126 @@
+package at.ac.tuwien.kr.alpha.grounder;
+
+import at.ac.tuwien.kr.alpha.common.ConstantTerm;
+import at.ac.tuwien.kr.alpha.common.FunctionTerm;
+import at.ac.tuwien.kr.alpha.common.Term;
+import at.ac.tuwien.kr.alpha.common.VariableTerm;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+
+public class Substitution {
+	private final HashMap<VariableTerm, Term> substitution;
+
+	public Substitution(HashMap<VariableTerm, Term> substitution) {
+		this.substitution = substitution;
+	}
+
+	public Substitution() {
+		this(new HashMap<>());
+	}
+
+	public Substitution(Substitution clone) {
+		this(new HashMap<>(clone.substitution));
+	}
+
+	public void replaceSubstitution(Substitution other) {
+		substitution.clear();
+		substitution.putAll(other.substitution);
+	}
```

Comment:
The main difference [to the previous version](https://github.com/AntoniusW/Alpha/pull/38/files/55317968094475905c0194fa76313f23fbd5a62a#diff-3bb08e391a45e7c42bba1e6471511498L665) is that `this.substitution` is final now, so the reference cannot be be changed. I will revert that.